### ISSUE TITLE
postcss Security Vulnerability

### ISSUE THREAD

**Issue by nzacca** [storybookjs/storybook#14883]
Hello!

Thought it would be good to open the discussion and see if someone might have an idea on how to proceed.

postcss versions less than 8.2.10 are vulnerable as per:  https://npmjs.com/advisories/1693 

It appears several projects under Storybook are using an old version of `css-loader` (< 5) or `postcss-flexbugs-fixes (< 5)` or `autoprefixer (< 10)` which have a dependency on `postcss` 7. For example:

Core-Server

- https://github.com/storybookjs/storybook/blob/master/lib/core-server/package.json#L66

Builder-Webpack4

- https://github.com/storybookjs/storybook/blob/master/lib/builder-webpack4/package.json#L76
- https://github.com/storybookjs/storybook/blob/master/lib/builder-webpack4/package.json#L82
- https://github.com/storybookjs/storybook/blob/master/lib/builder-webpack4/package.json#L93
- https://github.com/storybookjs/storybook/blob/master/lib/builder-webpack4/package.json#L94

Addon-postcss

- https://github.com/storybookjs/addon-postcss/blob/master/package.json#L46
- https://github.com/storybookjs/addon-postcss/blob/master/package.json#L47

Angular

- https://github.com/storybookjs/storybook/blob/master/app/angular/package.json#L50
- https://github.com/storybookjs/storybook/blob/master/app/angular/package.json#L54

These dependencies on older versions of postcss are preventing us from updating to a later version to address this vulnerability.

Is this something on the Storybook team's radar/planned for upcoming update?

**Comment by mayank99** [comment]
@storybook/addon-docs and @storybook/react also indirectly use postcss7

![image](https://user-images.githubusercontent.com/9084735/117859511-180add80-b25d-11eb-8aae-dceee9d81609.png)

**Comment by shilman** [comment]
You should be able to upgrade to the latest `postcss` using https://github.com/storybookjs/addon-postcss. It's a workaround until 7.0, when we plan to remove postcss support from the core and `addon-postcss` will be the recommended way to use `postcss` in Storybook.

**Comment by mayank99** [comment]
So we need to include the postcss addon for the security fix even if we are not using any postcss features?

**Comment by shilman** [comment]
If you're not using postcss does the security issue affect you?

**Comment by mayank99** [comment]
Yes, because postcss seems to be always included (indirectly) in multiple storybook packages.

**Comment by shilman** [comment]
If you don't execute the code is it actually a vulnerability? Installing `addon-postcss` is not going to magically uninstall this code either, but should guarantee that it doesn't execute.

**Comment by namirsab** [comment]
Hey everyone,
I'm getting vulnerabilities as well, and even though I'm considering installing the `addon-postcss` just to fix them, the main problem is that with `storybook` in the current state you can't run `npm audit fix` because I get the following error:

```
 npm audit fix
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! Found: postcss@7.0.35
npm ERR! node_modules/postcss
npm ERR!   postcss@"^7.0.35" from @storybook/builder-webpack4@6.2.9
npm ERR!   node_modules/@storybook/builder-webpack4
npm ERR!     @storybook/builder-webpack4@"6.2.9" from @storybook/addon-docs@6.2.9
npm ERR!     node_modules/@storybook/addon-docs
npm ERR!       @storybook/addon-docs@"6.2.9" from @storybook/addon-essentials@6.2.9
npm ERR!       node_modules/@storybook/addon-essentials
npm ERR!         dev @storybook/addon-essentials@"^6.2.9" from the root project
npm ERR!     @storybook/builder-webpack4@"6.2.9" from @storybook/core-server@6.2.9
npm ERR!     node_modules/@storybook/core-server
npm ERR!       @storybook/core-server@"6.2.9" from @storybook/core@6.2.9
npm ERR!       node_modules/@storybook/core
npm ERR!         @storybook/core@"6.2.9" from @storybook/addon-docs@6.2.9
npm ERR!         node_modules/@storybook/addon-docs
npm ERR!         1 more (@storybook/react)
npm ERR!   postcss@"^7.0.32" from css-loader@3.6.0
npm ERR!   node_modules/@storybook/builder-webpack4/node_modules/css-loader
npm ERR!     css-loader@"^3.6.0" from @storybook/builder-webpack4@6.2.9
npm ERR!     node_modules/@storybook/builder-webpack4
npm ERR!       @storybook/builder-webpack4@"6.2.9" from @storybook/addon-docs@6.2.9
npm ERR!       node_modules/@storybook/addon-docs
npm ERR!         @storybook/addon-docs@"6.2.9" from @storybook/addon-essentials@6.2.9
npm ERR!         node_modules/@storybook/addon-essentials
npm ERR!       1 more (@storybook/core-server)
npm ERR!   78 more (postcss-loader, css-loader, autoprefixer, ...)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer postcss@"^7.0.0 || ^8.0.1" from postcss-loader@4.2.0
npm ERR! node_modules/@storybook/builder-webpack4/node_modules/postcss-loader
npm ERR!   postcss-loader@"^4.2.0" from @storybook/builder-webpack4@6.2.9
npm ERR!   node_modules/@storybook/builder-webpack4
npm ERR!     @storybook/builder-webpack4@"6.2.9" from @storybook/addon-docs@6.2.9
npm ERR!     node_modules/@storybook/addon-docs
npm ERR!       @storybook/addon-docs@"6.2.9" from @storybook/addon-essentials@6.2.9
npm ERR!       node_modules/@storybook/addon-essentials
npm ERR!     1 more (@storybook/core-server)
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

**Comment by jetpack3331** [comment]
> Hey everyone,
> I'm getting vulnerabilities as well, and even though I'm considering installing the `addon-postcss` just to fix them, the main problem is that with `storybook` in the current state you can't run `npm audit fix` because I get the following error:
> 
> ```
>  npm audit fix
> npm ERR! code ERESOLVE
> npm ERR! ERESOLVE unable to resolve dependency tree
> npm ERR! 
> npm ERR! Found: postcss@7.0.35
> npm ERR! node_modules/postcss
> npm ERR!   postcss@"^7.0.35" from @storybook/builder-webpack4@6.2.9
> npm ERR!   node_modules/@storybook/builder-webpack4
> npm ERR!     @storybook/builder-webpack4@"6.2.9" from @storybook/addon-docs@6.2.9
> npm ERR!     node_modules/@storybook/addon-docs
> npm ERR!       @storybook/addon-docs@"6.2.9" from @storybook/addon-essentials@6.2.9
> npm ERR!       node_modules/@storybook/addon-essentials
> npm ERR!         dev @storybook/addon-essentials@"^6.2.9" from the root project
> npm ERR!     @storybook/builder-webpack4@"6.2.9" from @storybook/core-server@6.2.9
> npm ERR!     node_modules/@storybook/core-server
> npm ERR!       @storybook/core-server@"6.2.9" from @storybook/core@6.2.9
> npm ERR!       node_modules/@storybook/core
> npm ERR!         @storybook/core@"6.2.9" from @storybook/addon-docs@6.2.9
> npm ERR!         node_modules/@storybook/addon-docs
> npm ERR!         1 more (@storybook/react)
> npm ERR!   postcss@"^7.0.32" from css-loader@3.6.0
> npm ERR!   node_modules/@storybook/builder-webpack4/node_modules/css-loader
> npm ERR!     css-loader@"^3.6.0" from @storybook/builder-webpack4@6.2.9
> npm ERR!     node_modules/@storybook/builder-webpack4
> npm ERR!       @storybook/builder-webpack4@"6.2.9" from @storybook/addon-docs@6.2.9
> npm ERR!       node_modules/@storybook/addon-docs
> npm ERR!         @storybook/addon-docs@"6.2.9" from @storybook/addon-essentials@6.2.9
> npm ERR!         node_modules/@storybook/addon-essentials
> npm ERR!       1 more (@storybook/core-server)
> npm ERR!   78 more (postcss-loader, css-loader, autoprefixer, ...)
> npm ERR! 
> npm ERR! Could not resolve dependency:
> npm ERR! peer postcss@"^7.0.0 || ^8.0.1" from postcss-loader@4.2.0
> npm ERR! node_modules/@storybook/builder-webpack4/node_modules/postcss-loader
> npm ERR!   postcss-loader@"^4.2.0" from @storybook/builder-webpack4@6.2.9
> npm ERR!   node_modules/@storybook/builder-webpack4
> npm ERR!     @storybook/builder-webpack4@"6.2.9" from @storybook/addon-docs@6.2.9
> npm ERR!     node_modules/@storybook/addon-docs
> npm ERR!       @storybook/addon-docs@"6.2.9" from @storybook/addon-essentials@6.2.9
> npm ERR!       node_modules/@storybook/addon-essentials
> npm ERR!     1 more (@storybook/core-server)
> npm ERR! 
> npm ERR! Fix the upstream dependency conflict, or retry
> npm ERR! this command with --force, or --legacy-peer-deps
> npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
> ```
> 
> That's a real issue, as you can't even fix other vulnerabilities with the automatic tool because of this problem with peer dependencies.
> 
> or maybe this is a different issue?

Have the same problem...

**Comment by mayank99** [comment]
🎉 Postcss has backported its fix into 7.0.36! https://twitter.com/PostCSS/status/1403351574110511106

The advisory has been updated too: https://github.com/advisories/GHSA-hwj9-h5mp-3pm3

**Comment by shilman** [comment]
Ermahgerd!! I just released https://github.com/storybookjs/storybook/releases/tag/v6.3.0-rc.9 containing PR #15276 that references this issue. Upgrade today to the `@next` NPM tag to try it out!

```
npx sb upgrade --prerelease
```

Closing this issue. Please re-open if you think there's still more to do.

**Comment by pyymenta** [comment]
@shilman Thanks for this update. Unfortunately we still have the problem related in https://npmjs.com/advisories/1693.

To fix this, we must update postcss to 8.2.10 or later.
